arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -210,130 +209,203 @@ arangodb::Result fetchRevisions(
+          requestBuilder.add(toFetch[current + i].toValuePair(ridBuffer));
+          shoppingList.insert(toFetch[current + i]);
+        }
+        current += i;
       }
-    }
-    std::string request = requestBuilder->slice().toJson();
-
-    double t = TRI_microtime();
-    config.connection.lease(
-        [&](arangodb::httpclient::SimpleHttpClient* client) {
-          response.reset(client->retryRequest(arangodb::rest::RequestType::PUT,
-                                              url, request.data(),
-                                              request.size(), headers));
-        });
-    stats.waitedForDocs += TRI_microtime() - t;
-    ++stats.numDocsRequests;
-
-    if (arangodb::replutils::hasFailed(response.get())) {
-      return arangodb::replutils::buildHttpError(response.get(), url,
-                                                 config.connection);
-    }
 
-    arangodb::transaction::BuilderLeaser responseBuilder(&trx);
-    Result r = arangodb::replutils::parseResponse(*responseBuilder.get(),
-                                                  response.get());
-    if (r.fail()) {
-      return Result(
-          TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-          concatT("got invalid response from leader at ",
-                  config.leader.endpoint, url, ": ", r.errorMessage()));
-    }
-
-    VPackSlice docs = responseBuilder->slice();
-    if (!docs.isArray()) {
-      return Result(
-          TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-          concatT("got invalid response from leader at ",
-                  config.leader.endpoint, url, ": response is not an array"));
-    }
-
-    config.progress.set("applying documents by revision for collection '" +
-                        collection.name() + "'");
-
-    for (VPackSlice leaderDoc : VPackArrayIterator(docs)) {
-      if (!leaderDoc.isObject()) {
-        return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                      std::string("got invalid response from leader at ") +
-                          config.leader.endpoint + url +
-                          ": response document entry is not an object");
+      arangodb::network::RequestOptions reqOptions;
+      reqOptions.param("collection", leader)
+          .param("serverId", state.localServerIdString)
+          .param("batchId", std::to_string(config.batch.id));
+      reqOptions.database = config.vocbase.name();
+      reqOptions.timeout = arangodb::network::Timeout(25.0);
+      auto buffer = requestBuilder.steal();
+      auto f = arangodb::network::sendRequestRetry(
+          pool, config.leader.endpoint, arangodb::fuerte::RestVerb::Put, path,
+          std::move(*buffer), reqOptions);
+      futures.emplace_back(std::move(f));
+      shoppingLists.emplace_back(std::move(shoppingList));
+      ++stats.numDocsRequests;
+      LOG_TOPIC("eda42", DEBUG, arangodb::Logger::REPLICATION)
+          << "Have requested a chunk of 5000 revisions from "
```

Comment:
nit: the chunk can have a size < 5000 as well. In this case the log message will be (slightly) misleading. It's not important, just mentioning it here.
```suggestion
          << "Have requested a chunk of 5000 revisions from "
```

Reply:
Introduced a local counter. Good point.